### PR TITLE
Merge lab3

### PR DESCRIPTION
**Pull request by sding3**
David to review:
- [ ] `blaster.py` written by Shang
- [x] `blastee.py` written by Shang

Shang to review:
- [x] `middlebox.py` written by David

### CODE REVIEW

File: lab3/middlebox.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    out_port    - output port (EthAddr)
+    out_dst     - dest mac (EthAddr)
+'''
+def update_pkt(pkt, out_port, out_dst):
+    del pkt[0]
+    eth_head = Ethernet(src=out_port, dst=out_dst, ethertype=EtherType.IPv4)
+    pkt.prepend_header(eth_head)
+
+    #Decrement TTL
+    pkt[IPv4].ttl -= 1
```
Comment (sding3) [inline lab3/middlebox.py:72]: There needs to be additional handling to check if the ttl, after the decrement, is less than or equal to zero, to determine whether to forward or drop the packet. Without this additional handling, the middlebox crashes when a packet arrives with TTL <= 1.

File: lab3/middlebox.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    config = load_file("./middlebox_params.txt")
+
+    random_seed = config['s']
+    random.seed(random_seed) #Extract random seed from params file
+
+    while True:
+        gotpkt = True
+        try:
+            timestamp,dev,pkt = net.recv_packet()
```
Comment (sding3) [inline lab3/middlebox.py]: It doesn't appear that `timestamp` is used anywhere below. If that's the case, it should be replaced with `_`

File: lab3/middlebox.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    mymacs = [intf.ethaddr for intf in my_intf]
+    myips = [intf.ipaddr for intf in my_intf]
+
+    config = load_file("./middlebox_params.txt")
+
+    random_seed = config['s']
+    random.seed(random_seed) #Extract random seed from params file
+
+    while True:
+        gotpkt = True
```
Comment (sding3) [inline lab3/middlebox.py]: We can obviate the need for this `gotpkt` state tracking variable if you replace the `except NoPackets:` block with this:

```

        except NoPackets:
            log_debug("No packets available in recv_packet")
            continue 
```

With the current code, even in the no-packet case, line through 103 through 137 will run. These lines shouldn't be run in the no-packet case.

File: lab3/middlebox.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+"""
+David to implement
+"""
+
+from switchyard.lib.address import *
+from switchyard.lib.packet import *
+from switchyard.lib.packet.util import *
+from switchyard.lib.userlib import *
+from threading import *
```
Comment (sding3) [inline lab3/middlebox.py]: I think this `threading import` was provided in the course provided starting point, but it doesn't appear that we use anything from it. So this should be removed.

**Review by dmbillmireUW — APPROVED** [review]

**Pull request merged**